src/Map.php

```diff
@@ -399,14 +487,22 @@ public function pairs(): Sequence
      */
     public function put($key, $value)
     {
-        $pair = $this->lookupKey($key);
-
-        if ($pair) {
-            $pair->value = $value;
+        $pairRef = $this->lookupKey($key, $hash);
 
+        if ($pairRef) {
+            $pairRef->pair->value = $value;
         } else {
             $this->checkCapacity();
-            $this->pairs[] = new Pair($key, $value);
+            $pair = new Pair($key, $value);
+            $pairIndex = $this->nextPairIndex++;
```

Comment:
is there a reason we cant just `$this->pairRefs[] = $pairRef?

Reply:
We need the index to go into the `PairRef` constructor so that we can keep it up to date when the map order changes, e.g., after a reverse or sort. We can't get the index from `count($this->pairs)` because the array can have holes in it from deletes.